### PR TITLE
[Enhancement] Edit page title from page editor

### LINKED ISSUE #2125: Make page title editable from the editor
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

Currently, our page title is not editable from the editor. (`Employees` from the below screenshots)
A user can edit it from the page.yaml. This task is about editing it from the Toolpad UI.

Screenshot 2023-06-05 at 1 16 51 PM

Screenshot 2023-06-05 at 1 20 43 PM

### Examples 🌈

_No response_

### Motivation 🔦

_No response_

### PR DESCRIPTION
**Pull request by Kirera-Wainaina**
This PR adds a Textfield to the page panel and allows a user to change their page title. It closes #2125.

- [x] I've read and followed the [contributing guide](https://github.com/mui/mui-toolpad/blob/master/CONTRIBUTING.md#sending-a-pull-request) on how to create great pull requests.
- [ ] I've updated the relevant documentation for any new or updated feature
- [x] I've linked relevant GitHub issue with "Closes #<issue id>"
- [x] I've added a visual demonstration under the form of a screenshot or video
image

### CODE REVIEW

File: packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    (event: React.ChangeEvent<HTMLInputElement>) => setPageTitleInput(event.target.value),
+    [],
+  );
+
+  const handleCommit = React.useCallback(() => {
+    if (!isValid) {
+      return;
+    }
+    domApi.update((appDom: AppDom) => {
+      return update(appDom, {
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx:26]: You can use [`setNamespacedProp`](https://github.com/mui/mui-toolpad/blob/eacf6f9cdbce81868c44d38055e398a6980d7022/packages/toolpad-app/src/appDom/index.ts#L713C35-L713C35) for this.

File: packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  const handleKeyPress = React.useCallback(
+    (event: React.KeyboardEvent<HTMLDivElement>) => {
+      if (event.code === 'Enter') {
+        handleCommit();
+      }
+    },
+    [handleCommit],
+  );
+
+  React.useEffect(() => {
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx:46]: No need for creating new state and an effect here. You can just derive state as follows:

```tsx
const isValid = !!pageTitleInput
```

Personally I'd rather solve this by creating the error first as

```tsx
const inputError = React.useMemo(() => {
  if (!pageTitleInput) {
    return 'Input require'
  }
  if (pageTitleInput.length > MAX_TITLE_SIZE) {
    return 'Page title too long'
  }
  // ...
  return null
}, [pageTitleInput])

const isValid = !inputError
```

But I'm wondering, wouldn't this make it impossible to remove a title from a page?

Comment (Kirera-Wainaina) [reply packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx:46]: If we link `validateInput` with `handleCommit`. It should work.

File: packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx (hunk 1 above)
Comment (Janpot) [reply packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx:26]: @Kirera-Wainaina Sorry, I meant [`setNodeNamespacedProp`](https://github.com/mui/mui-toolpad/blob/eacf6f9cdbce81868c44d38055e398a6980d7022/packages/toolpad-app/src/appDom/index.ts#L739)

**Comment by Kirera-Wainaina** [comment]
As for `MAX_TITLE_SIZE`, the [recommended is 64 characters](https://www.w3.org/Provider/Style/TITLE.html#:~:text=The%20title%20should%20ideally%20be,there%20is%20only%20limited%20room.) and since there is ` | Toolpad Editor` as the suffix, then we can set max to 47 characters. What do you think?

**Comment by Janpot** [comment]
> What do you think?

On second thought, I don't think we should add a hard limit at all. Users can fill in the file whatever they want anyway. We should rather focus on proper truncation when we use the title when rendering the page.

**Comment by Kirera-Wainaina** [comment]
I've made the changes we discussed. Let me know if there is something else.

File: packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  const handleKeyPress = React.useCallback(
+    (event: React.KeyboardEvent<HTMLDivElement>) => {
+      if (event.code === 'Enter') {
+        handleCommit();
+      }
+    },
+    [handleCommit],
+  );
+
+  const validateInput = (input: string) => {
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx:34]: This can be moved outside of the components

**Review by Janpot — CHANGES_REQUESTED** [review]
👍 Looking great. Just one small stylistic change and then we can merge

**Comment by Kirera-Wainaina** [comment]
Have a look.

**Review by Janpot — APPROVED** [review]
thank you!

**Pull request merged**